### PR TITLE
Changed choco installation according previously finished work

### LINKED ISSUE #48: 💡 [REQUEST] - properly filter choco installation when package is needed to be upgraded instead of reinstall
### Reference Issues

_No response_

### Summary

Currently choco package is always performing command choco install --force even when it's needed to be updating.it's nessesary to create difference whenewer we need to install or upgrade package

### Basic Example

1) we have already installed package:
 system detects package version and if it differs from current one launces upgrade
2) we have no such installed package:
System just installes the package
3) system has installed package with the Same version as recieved:
Package should be rejected as already installed, result is positive

### Drawbacks

-

### Unresolved questions

_No response_

### PR DESCRIPTION
**Pull request by SteamDragon**
Fixes #48 

## Proposed Changes
For choco: 
  - Installation is now replaced with Upgrade when package is already installed
  - All PS executes is now replaced with correct process calls
  - Added script to manually test package of nuget
  - Fixed uninstall key to correctly upgrade

### CODE REVIEW

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 1]
```diff
@@ -55,31 +68,43 @@ public bool IsPackageInstalled(Package package)
...
+            Process p = new Process();
+            p.StartInfo.FileName = "choco.exe";
+            p.StartInfo.Arguments = "list -l";
+            p.StartInfo.RedirectStandardOutput = true;
+            p.StartInfo.UseShellExecute = false;
+            p.Start();
+            p.WaitForExit();
+            productCodes.Clear();
+            productCodes.AddRange(p.StandardOutput.ReadToEnd()
+                .Split(new[] { Environment.NewLine },
```
Comment (vchaplinski) [inline Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:80]: Are you sure such split does the right job?
Sometimes (depends on application) the output may contain CR+LF sequence at the end of each line.
In this case the elements would still contain CR at the end which is not what we expect.

I would implement it in more clear and straight-forward way using StandardOutput.ReadLine() method flitering out the lines which string.IsNullOrWhiteSpace()

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 2]
```diff
@@ -55,31 +68,43 @@ public bool IsPackageInstalled(Package package)
...
-            using (var ps = PowerShell.Create())
+            try
+            {
+                Process p = new Process();
+                p.StartInfo.FileName = "choco.exe";
+                p.StartInfo.Arguments = "--version";
+                p.StartInfo.RedirectStandardOutput = true;
+                p.StartInfo.UseShellExecute = false;
+                p.Start();
+                p.WaitForExit();
```
Comment (vchaplinski) [inline Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:102]: I don't think we need to wait for exit.
If choco is not installed the p.Start() will throw exception immediately (please check it anyway).

File: package/CreateChocoPackage.ps1 [hunk 3]
```diff
@@ -0,0 +1,10 @@
+$content=(Get-Content -path tools\chocolateyinstall.ps1 -Raw)
```
Comment (vchaplinski) [inline package/CreateChocoPackage.ps1:1]: Is it possible to add a short description in the comments at the top of this script, with some instruction of usage?

**Review by vchaplinski — CHANGES_REQUESTED** [review]

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 4]
```diff
@@ -55,31 +68,43 @@ public bool IsPackageInstalled(Package package)
 
         public void Refresh()
         {
-            using (var ps = PowerShell.Create())
```
[the remark this answers is not in the document]
Comment (SteamDragon) [reply Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:58]: Done

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 5]
```diff
@@ -36,16 +37,28 @@ public Process StartInstallationProcess(Package package)
...
-                                    "-y --force --no-progress";
+                                    "-y --no-progress";
             p.StartInfo.RedirectStandardOutput = true;
             p.StartInfo.UseShellExecute = false;
             p.Start();
 
             return p;
         }
 
+        private string GetInstallationType(Package package)
```
[the remark this answers is not in the document]
Comment (SteamDragon) [reply Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:51]: Changed

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 6]
```diff
@@ -36,16 +37,28 @@ public Process StartInstallationProcess(Package package)
...
+        {
+            string packageInstallationType = "install";
+            if (productCodes.Any(item => item.StartsWith(package.ProductCode.Split(' ').First())))
```
[the remark this answers is not in the document]
Comment (SteamDragon) [reply Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:54]: Changed

File: package/CreateChocoPackage.ps1 (hunk 3 above)
Comment (SteamDragon) [reply package/CreateChocoPackage.ps1]: Added info

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs [hunk 7]
```diff
@@ -55,31 +62,44 @@ public bool IsPackageInstalled(Package package)
...
+            p.WaitForExit();
+            productCodes.Clear();
+            StreamReader standardOutput = p.StandardOutput;
+            while (!standardOutput.EndOfStream)
             {
-                const string psCommand = @"choco list -li";
-                ps.AddScript(psCommand);
-                productCodes.Clear();
-                productCodes.AddRange(ps.Invoke<string>());
+                productCodes.Add(standardOutput.ReadLine());
```
Comment (vchaplinski) [inline Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs:76]: After attentive look into the Choco output...
Seems like 1st line shall be skipped.
Also the last line is empty - we don't want it.
And the line above it (last visible one) contains statistics. Also must be excluded.

**Review by vchaplinski — CHANGES_REQUESTED** [review]

File: Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs (hunk 7 above)
Comment (SteamDragon) [reply Up2dateService/Up2dateService/Installers/Choco/ChocoInstaller.cs]: Lines ignored

**Pull request merged**